Expose a prop allowing users to specify a custom fallback image

jingdid opened the pull request:
This change allows a user to specify a custom fallback image by setting a `placeholderImg` prop  to a url like so:
`<Carousel images={images} placeholderImg="https://placekitten.com/400/230" />`

If the placeholder image url provided is invalid, the repository's default placeholder image is used instead.

Addresses issue #86

yifaneye committed:
Update version number (#85)

Commit:
Expose a prop allowing users to specify a custom fallback image

yifaneye commented:
Hi @jingdid ! Thank you very much! I will review it today!

yifaneye commented on yarn.lock:
Would you like to tell me what is the reason for the changes here? Could you please keep the yarn lockfile unchanged?

jingdid replied:
Apologies, I did not notice I had changed the yarn lockfile. I think this is just a consequence of my trying to get the demo to run locally. I don't think any of my changes should actually require updates to the lockfile - I will remove this when I next update this pull request.

yifaneye commented on src/components/Image/Image.js:
```diff
@@ -33,15 +33,27 @@ const LazyLoadedImage = (props) => {
+  const [isLoading, setIsLoading] = useState(true);
+  const [isValidFallback, setIsValidFallback] = useState(false);
+  useEffect(() => {
+    fetch(props.placeholderImg).then(res => {
```
Does this implementation support local file paths?

jingdid replied:
Tested it with a local file and it should support them.

yifaneye commented on src/components/Carousel/props.js:
```diff
@@ -48,6 +48,7 @@ export const propTypes = {
   hasDotButtonsAtMax: largeWidgetPositions,
   hasCaptionsAtMax: largeWidgetPositions,
   hasThumbnailsAtMax: PropTypes.bool,
+  placeholderImg: PropTypes.string,
```
I'd suggest renaming it to `fallbackImg`.

yifaneye replied:
And also in other props.

jingdid replied:
Done.

yifaneye commented on src/components/Image/Image.js:
```diff
@@ -33,15 +33,27 @@ const LazyLoadedImage = (props) => {
+    fetch(props.placeholderImg).then(res => {
+      setIsValidFallback(res.status === 200);
+      setIsLoading(false);
+    }).catch((_reason) => {
```
Here, `finally` statement can be used to combine `setIsLoading(false);`.

jingdid replied:
This block has now been moved out of this file, and is used in such a way now that `finally` should no longer be needed.

yifaneye commented on src/components/Image/Image.js:
```diff
@@ -33,15 +33,27 @@ const LazyLoadedImage = (props) => {
+    }).catch((_reason) => {
+      setIsLoading(false);
+    })
+  }, []);
```
Here is an error message: 
> ESLint: React Hook useEffect has a missing dependency: 'props.placeholderImg'. Either include it or remove the dependency array.(react-hooks/exhaustive-deps)

jingdid replied:
Done.

yifaneye commented on src/components/Image/Image.js:
```diff
@@ -33,15 +33,27 @@ const LazyLoadedImage = (props) => {
+  const [isValidFallback, setIsValidFallback] = useState(false);
+  useEffect(() => {
+    fetch(props.placeholderImg).then(res => {
+      setIsValidFallback(res.status === 200);
```
Could it be "HTTP 304 Not Modified" in some cases?

jingdid replied:
I've changed it to check for `res.ok` instead now, which should address this issue.

yifaneye commented on src/components/Image/Image.js:
```diff
@@ -81,11 +93,28 @@ export const Image = (props) => {
+    }).catch((_reason) => {
+      setIsLoading(false);
+    })
+  }, []);
```
Here is an error message:
`ESLint: React Hook useEffect has a missing dependency: 'props.placeholderImg'. Either include it or remove the dependency array.(react-hooks/exhaustive-deps)`

jingdid replied:
Done.

yifaneye commented on src/components/Image/Image.js:
```diff
@@ -33,15 +33,27 @@ const LazyLoadedImage = (props) => {
 
+  // This block checks whether the user specified fallback image actually works or not
+  const [isLoading, setIsLoading] = useState(true);
+  const [isValidFallback, setIsValidFallback] = useState(false);
```
What's the reason for handling the error of the fallback image? The fallback image is likely to be a static image, which is quite stable. The user can handle it using their code if they want to, isn't it?

yifaneye replied:
We can simply assign the fallback image in the `handleError` function.

jingdid replied:
This was intended to catch cases where a user provides an invalid/broken url for the fallback image - in that event, I wanted to make sure that the carousel would end up falling back onto the built-in placeholder again.

yifaneye reviewed (changes requested):
Thank you for your contribution! Your code looks good. But I left some comments.

yifaneye reviewed (changes requested):
Thank you for your contribution! Your code looks good. But I left some comments.

jingdid committed:
Address reviewer feedback and improve fallback image implementation

- Refactor so that fallback image validity is only check once rather
than making a request for every image in the carousel
- Allow thumbnails to also use fallback image
- Gracefully use default placeholder if fallback image prop is not
specified

jingdid commented:
@yifaneye Welcome back! Apologies for the bother, but any chance you could re-review this pull request? I've since addressed your comments while also improving the changed behavior.
